File: Rohit Singhal/LinkedLists/PalindromeLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    
     public ListNode reverse(ListNode head)
     {
         if(head==null || head.next==null)
         {
             return head;
         }
         ListNode dummy=null;
         ListNode next;
         while(head!=null)
         {
             next=head.next;
             head.next=dummy;
             dummy=head;
             head=next;
         }
         return dummy;
     }
    
    
    public boolean isPalindrome(ListNode head) {
        if(head==null || head.next==null)
        {
            return true;
        }
        ListNode slow=head;
        ListNode fast=head;
        while(fast.next!=null && fast.next.next!=null)
        {
            slow=slow.next;
            fast=fast.next.next;
        }
        slow.next=reverse(slow.next);
        slow=slow.next;
        ListNode temp=head;
        while(slow!=null)
        {
            if(slow.val!=temp.val)
            {
                return false;
            }
            slow=slow.next;
            temp=temp.next;
        }
        return true;

        
    }
}

File: Shreya/LinkedList,Queues,Stacks/Palindrome Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (head->next == NULL)
        {
            return true;
        }
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast!=NULL && fast->next!=NULL)
        {
            slow = slow->next;
            fast = fast->next->next;
        }
        slow = reversed(slow);
        fast = head;
        while(slow!=NULL)
        {
            if(slow->val!=fast->val)
            {
                return false;
            }
            slow=slow->next;
            fast=fast->next;
        }
        return true;
    }
    public:
    ListNode* reversed(ListNode* slow)
    {
        ListNode* prev = NULL;
        ListNode* temp = slow;
        while(temp!=NULL)
        {
            ListNode* curr = temp->next;
            temp->next = prev;
            prev = temp;
            temp = curr;
        }
        slow = prev;
        return slow;
    }
};